ShoppingM8-be/src/main/java/edu/pw/shoppingm8/product/api/ProductController.java

```diff
@@ -81,10 +90,11 @@
         @ApiResponse(code = 403, message = "If user does not have access to the list"),
         @ApiResponse(code = 404, message = "If invalid id was provided")})
     @PatchMapping("{id}")
-    public ResponseEntity<Void> patchProduct(
-        @PathVariable Long listId, @PathVariable Long id, @Valid @RequestBody ProductPatchRequestDto productDto) {
-        productService.updateProduct(listId, id, productDto);
-        return ResponseEntity.ok().build();
+    public ResponseEntity<ProductResponseDto> updateProduct(
+            @PathVariable Long listId, @PathVariable Long id, @Valid @RequestBody ProductPatchRequestDto productDto) {
+        List list = listService.getList(listId);
+        listService.checkIfUserHasAccessTo(list);
```

Comment:
It depends how you look at it. Some will say that controller layer should only be a clean proxy between services and the outside world and they should not operate on domain objects, throwing them between services. Others will argue that controller layer is allowed to use whole service layer this way, because it doesn't violate architecture priciples and it allows better separation of services (also reduces possible circular dependencies). 

For this particular line I would say I partially agree. IMHO good approach to that is by validation through spring aspects, because you have both separated business logic, hidden from API layer and well separated services. I didn't do it here to use the code that already existed. You can look at my other project, BoardUp, and see if that way would be better.  If so, I can quickly make those changes.